Add get_job function

bdice opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->

## Description
When the current working directory is a signac job directory, it would be nice to be able to quickly identify that job in a similar way to `signac.get_project()`. This implementation

## Motivation and Context
Resolves #108. Sometimes I navigate the filesystem to a job directory (e.g. `cd workspace/abc123`) so I can access files directly. However, it is tough to open a Python REPL and get to this job without copy-pasting the job id.

## Types of Changes
<!-- Please select all items that apply either now or after creating the pull request: -->
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

This pull request is based on
- [ ] *master*, because it is a bug fix or an update to the documentation.
- [x] *develop*, because it introduces a new feature.

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:
<!-- Please select all items that apply either now or after creating the pull request. -->
<!-- If you are unsure about any of these items, do not hesitate to ask! -->
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

csadorf commented on signac/contrib/project.py:
```diff
@@ -1524,6 +1524,36 @@ def get_project(cls, root=None):
 
+    @classmethod
+    def get_job(cls, root=None):
+        """From a job workspace directory or subdirectory, get the job.
```
I think this needs to be clarified.

bdice replied:
I changed this.

csadorf commented on signac/contrib/project.py:
```diff
@@ -1524,6 +1524,36 @@ def get_project(cls, root=None):
+        if root is None:
+            root = os.getcwd()
+        results = re.findall(JOB_ID_REGEX, root)
+        if len(results) > 0:
```
Shouldn't the length of `results` be exactly one? What does it mean if it is larger?

bdice replied:
We're doing a regex match on the full path. If you're in a directory like `my_project/workspace/JOB_ID/my_subproject/workspace/SUBJOB_ID`, then we take `SUBJOB_ID` (the last match). Then this will be a job belonging to the project `my_subproject` found from that same directory.

csadorf commented on signac/contrib/project.py:
```diff
@@ -1827,3 +1857,16 @@ def get_project(root=None):
+
+
+def get_job(root=None):
+    """From a job workspace directory, locate the closest parent project
```
I think this is supposed to express "when the current working directory is a workspace directory" or so, but I don't think this is clear with the current description.

bdice replied:
I changed this.

csadorf commented on signac/contrib/project.py:
```diff
@@ -1827,3 +1857,16 @@ def get_project(root=None):
+def get_job(root=None):
+    """From a job workspace directory, locate the closest parent project
+    and return this job.
+
```
Can we maybe add a simple example on how this function is to be used here?

bdice replied:
Added an example.

csadorf commented on tests/test_project.py:
```diff
@@ -1827,6 +1827,58 @@ def check_root(root=None):
+        # Test case: The job workspace dir is a sub-directory of the project root dir.
+        pass
+
+    def test_get_job_not_subdir_of_project(self):
```
I think I understand this unit test that still needs to implemented, but I forgot what the one above is supposed to test. @bdice Do you remember?

bdice replied:
Two (three?) tests are unimplemented:
```python
    def test_get_job_subdir_of_project(self):
        # Test case: The job workspace dir is a sub-directory of the project root dir.
        pass
```
This description doesn't make sense to me. I would like to test the case where we have a project with a job, and the current working directory is a subdirectory of the job workspace.
```python
    def test_get_job_not_subdir_of_project(self):
        # Test case: The job workspace dir is not a sub-directory of project root dir.
        # Test case: The job is not actually part of the project that we identified for this job.
        pass
```
I think the first case would be like making a symlink to `my_project/workspace/abc123` and placing that symlink outside of the `my_project` directory. I expect this to fail.

I think the second case is similar to the first, but placing that symlink in an unrelated project `other_project` directory. This should also fail unless that job symlink is placed in the workspace, `other_project/workspace`, in which case it is actually "in the `other_project`." Not sure what to test there.

If you want me to implement these, ping me.

bdice commented:
@csadorf Ready for additional review.

csadorf commented:
I'm going to rebase this branch on develop.

csadorf commented:
> I'm going to rebase this branch on develop.

Just kidding, I merged it.

bdice commented:
@csadorf I have implemented three additional tests for various cases and added more comments to the logic of `get_job`. This addresses all of the cases I can think of. I'm ready to call this done when tests pass.

csadorf commented on signac/contrib/project.py:
```diff
@@ -1531,6 +1531,49 @@ def get_project(cls, root=None, search=True):
+    @classmethod
+    def get_job(cls, root=None):
+        """Find a Job in or above the current working directory (or
+        provided path).
```
The first paragraph in a doc-string should only span one line: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

csadorf commented on signac/contrib/project.py:
```diff
@@ -1531,6 +1531,49 @@ def get_project(cls, root=None, search=True):
+
+        # Find a project *above* the root directory (if the provided root
+        # contains a project, we must search from its parent)
+        project = cls.get_project(os.path.join(job_root, os.pardir))
```
I don't quite understand the comment here. It seems like we are *always* searching the parent directory, but the comment implies that we do this only sometimes.

csadorf commented on signac/contrib/project.py:
```diff
@@ -1531,6 +1531,49 @@ def get_project(cls, root=None, search=True):
+        try:
+            job = project.open_job(id=job_id)
+        except KeyError:
+            raise
```
Since we are just re-raising we can completely omit the above two lines.

csadorf commented on signac/contrib/project.py:
```diff
@@ -1531,6 +1531,49 @@ def get_project(cls, root=None, search=True):
+            if job in project:
+                return job
+            else:
+                raise LookupError("Unable to find job id {} in project '{}'.".format(
```
I think we can just rely on the exception raised by the `open_job()` function in line 1567, can we not?

csadorf commented on signac/contrib/project.py:
```diff
@@ -1836,6 +1879,30 @@ def get_project(root=None, search=True):
+def get_job(root=None):
+    """When the current working directory or provided root directory is a job
+    workspace directory (or subdirectory thereof), locate the closest parent
+    project and return this job.
```
Same comment as above, the first paragraph in a multi-line doc string should span only one line.

csadorf commented on tests/test_project.py:
```diff
@@ -1827,19 +1841,90 @@ def check_root(root=None):
+        with job:
+            os.mkdir('test_subdir')
+            self.assertEqual(project.get_job('test_subdir'), job)
+            self.assertEqual(signac.get_job('test_subdir'), job)
```
Can we also test this without switching into the job>

csadorf commented on tests/test_project.py:
```diff
@@ -1827,19 +1841,90 @@ def check_root(root=None):
+            nestedproject.open_job({'b': 2}).init()
+            os.mkdir('test_subdir')
+            self.assertEqual(project.get_job('test_subdir'), job)
+            self.assertEqual(signac.get_job('test_subdir'), job)
```
Same as above, this should be tested also without switching into the job context.

bdice commented:
@csadorf I addressed your comments. Should be ready to go with your approval.

bdice commented:
@csadorf There are H5Store / numpy related failures on pypy2/3. Doesn't seem related to this PR.

csadorf commented:
> @csadorf There are H5Store / numpy related failures on pypy2/3. Doesn't seem related to this PR.

Not sure how this is possible... 😕  I'll have a look.

csadorf commented:
@bdice I've added an entry to the changelog. I'm ok with updating framework documentation as part of some general overhaul.
